src/Data/Text/Builder/Linear/Dec/Unbounded.hs

```diff
@@ -0,0 +1,332 @@
+    {-# INLINE prependTiny #-}
+    prependTiny ∷ Bool → DigitsWriter s
+    prependTiny !high !o1 !n# =
+      case BN.bigNatQuotRemWord# n# poweredBase of
+        (# q, r #) → do
+          !o2 ← unsafePrependWordDec marr (I# o1) (W# r)
+          if high && BN.bigNatIsZero q
+            then pure o2
+            else do
+              let !o3 = I# o1 - (fromIntegral (W# power) - 1)
+              padWithZeros marr o3 (o2 - o3)
+              !o4 ← unsafePrependWordDec marr (o3 - 1) (BN.bigNatToWord q)
+              if high
+                then pure o4
+                else do
+                  let !o5 = o3 - fromIntegral (W# power)
+                  padWithZeros marr o5 (o4 - o5)
+                  pure o5
+
+--------------------------------------------------------------------------------
+-- Prepend word
+--------------------------------------------------------------------------------
+
+unsafePrependWordDec ∷ ∀ s. A.MArray s → Int → Word → ST s Int
+unsafePrependWordDec = f
+  where
+    f marr !o !k
+      | k >= 10 = do
+          let (q, r) = quotRem100 k
+          A.copyFromPointer marr (o - 1) (digits `plusPtr` (fromIntegral r `shiftL` 1)) 2
+          if k < 100 then pure (o - 1) else f marr (o - 2) q
+      | otherwise = do
+          A.unsafeWrite marr o (fromIntegral (0x30 + k))
+          pure o
+
+--------------------------------------------------------------------------------
+-- Utils
+--------------------------------------------------------------------------------
+
+maxIntegerDecLen ∷ Integer → Int
+maxIntegerDecLen a = case a of
+  I.IS i# → maxDecLen (I# i#)
+  I.IP n# → maxBitNatDecLen n#
+  I.IN n# → 1 + maxBitNatDecLen n#
+{-# INLINEABLE maxIntegerDecLen #-}
+
+-- | ceiling (fbs a * logBase 10 2) < ceiling (fbs a * 5 / 16) < 1 + floor (fbs a * 5 / 16)
+--
+-- We approximate @fbs a@ to @bigNatSize a * word_size@.
+maxBitNatDecLen ∷ BN.BigNat# → Int
+maxBitNatDecLen n# =
+  -- This can overflow in theory, but in practice it would overflow for a BigNat#
+  -- of at least:
+  --
+  -- • On 32 bits platform: 0.8 GiB, out of max 4 GiB RAM
+  --   → BN.bigNatSize n# = 214748364 = min
+  --       (toInteger (maxBound @Word64 `div` fromIntegral (finiteBitSize @Word32 0 * 5)))
+  --       (((toInteger (maxBound @Int32) - 1) * 16) `div` fromIntegral (finiteBitSize @Word32 0 * 5))
+  -- • On 64 bits platform: 461 PiB
+  --   → BN.bigNatSize n# = 57646075230342348 = min
+  --       (toInteger (maxBound @Word64 `div` fromIntegral (finiteBitSize @Word64 0 * 5)))
+  --       (((toInteger (maxBound @Int64) - 1) * 16) `div` fromIntegral (finiteBitSize @Word64 0 * 5))
+  --
+  -- These thresholds are too big to be realistic (32 bits: 20% of RAM, 64 bits:
+  -- integer size in petabytes), so it seems perfectly reasonable to have no
+  -- special handling of overflow here.
+  1
+    + fromIntegral @Word64
+      ( (fromIntegral (BN.bigNatSize n#) * fromIntegral (finiteBitSize @Word 0) * 5)
+          `shiftR` 4
```

Comment:
Can we divide by 16 first and multiply by 5 second? It would allow to avoid any overflows I imagine. Might require something larger than `1 + ...`, but if we are in a business of rendering large nats a few extra bytes would not matter.

Reply:
Reworked: on common arch (32 and 64 bits) `finiteBitSize @Word 0` is a multiple of 16 and is `>= 29` by the Haskell report. So we can indeed simply this and it will compile to core with just a multiplication (10 or 20 respectively). I kept the original implementation for other (unlikely) arch.